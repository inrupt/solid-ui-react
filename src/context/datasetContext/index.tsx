//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import React, { createContext, ReactElement, useState, useEffect } from "react";
import {
  SolidDataset,
  WithResourceInfo,
  UrlString,
} from "@inrupt/solid-client";

import useDataset from "../../hooks/useDataset";

export interface IDatasetContext {
  solidDataset: SolidDataset | (SolidDataset & WithResourceInfo) | undefined;
  setDataset: (solidDataset: SolidDataset) => void;
}

const DatasetContext = createContext<IDatasetContext>({
  solidDataset: undefined,
  setDataset: () => {},
});

export default DatasetContext;

export interface IDatasetProvider {
  children: React.ReactNode;
  /**
   * @deprecated `loading` is deprecated. It can still be used but may be removed in a future major release.`loadingComponent` should be used instead.
   */
  loading?: React.ReactNode;
  loadingComponent?: React.ComponentType | null;
  onError?(error: Error): void | null;
  solidDataset?: SolidDataset | (SolidDataset & WithResourceInfo);
  datasetUrl?: UrlString | string;
}

export type RequireProperty<T, Prop extends keyof T> = T & {
  [key in Prop]-?: T[key];
};

export type RequireDatasetOrDatasetUrl =
  | RequireProperty<IDatasetProvider, "solidDataset">
  | RequireProperty<IDatasetProvider, "datasetUrl">;

/**
 * Used to provide a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) to child components through context, as used by various provided components and the useDataset hook.
 */
export const DatasetProvider = ({
  children,
  onError,
  solidDataset: propDataset,
  datasetUrl,
  loading,
  loadingComponent: LoadingComponent,
}: RequireDatasetOrDatasetUrl): ReactElement => {
  const { dataset, error } = useDataset(datasetUrl);

  if (error && onError) {
    onError(error);
  }

  const datasetToUse = propDataset ?? dataset;

  // Provide a setDataset function so that child components can update.
  const [stateDataset, setDataset] = useState<
    SolidDataset | (SolidDataset & WithResourceInfo) | undefined
  >(datasetToUse);

  // If the dataset is asynchronously loaded, make sure to set the new state value.
  useEffect(() => {
    setDataset(datasetToUse);
  }, [datasetToUse]);

  let loader: JSX.Element | React.ReactNode | null = (LoadingComponent && (
    <LoadingComponent />
  )) ||
    loading || <span>Fetching data...</span>;

  if (LoadingComponent === null) {
    loader = null;
  }

  return (
    <DatasetContext.Provider value={{ solidDataset: stateDataset, setDataset }}>
      {stateDataset ? children : loader}
    </DatasetContext.Provider>
  );
};
