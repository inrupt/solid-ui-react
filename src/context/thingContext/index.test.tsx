/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import * as React from "react";
import { RenderResult, render } from "@testing-library/react";
import * as SolidFns from "@inrupt/solid-client";
import { ThingContext, ThingProvider } from "./index";
import { DatasetProvider } from "../datasetContext/index";

let documentBody: RenderResult;

const mockPredicate = `http://xmlns.com/foaf/0.1/nick`;
const mockNick = "test nick value";

const mockThing = SolidFns.addStringNoLocale(
  SolidFns.createThing(),
  mockPredicate,
  mockNick
);

// const mockDataSet = SolidFns.setThing(SolidFns.createLitDataset(), mockThing);
const mockDataSetWithResourceInfo = SolidFns.setThing(
  SolidFns.createLitDataset() as any,
  mockThing
);

// TODO: refactor this once ticket SDK-1157 has been done
mockDataSetWithResourceInfo.internal_resourceInfo = {};
mockDataSetWithResourceInfo.internal_resourceInfo.fetchedFrom =
  "https://some-interesting-value.com";

function ExampleComponentWithThing(): React.ReactElement {
  const [property, setProperty] = React.useState<string>(
    "fetching in progress"
  );
  const thingContext = React.useContext(ThingContext);
  const { thing } = thingContext;

  React.useEffect(() => {
    if (thing) {
      const fetchedProperty = SolidFns.getStringUnlocalizedOne(
        thing,
        "http://xmlns.com/foaf/0.1/name"
      );
      if (fetchedProperty) {
        setProperty(fetchedProperty);
      }
    }
  }, [thing]);

  return (
    <div>
      <h2>{property}</h2>
    </div>
  );
}

function ExampleComponentWithThingUrl(): React.ReactElement {
  const examplePredicate = "http://xmlns.com/foaf/0.1/nick";
  const [property, setProperty] = React.useState<string>(
    "fetching in progress"
  );

  const thingContext = React.useContext(ThingContext);
  const { thing } = thingContext;

  React.useEffect(() => {
    if (thing) {
      const fetchedProperty = SolidFns.getStringUnlocalizedOne(
        thing,
        examplePredicate
      );
      if (fetchedProperty) {
        setProperty(fetchedProperty);
      }
    }
  }, [examplePredicate, thing]);

  return (
    <div>
      <h2>{property}</h2>
    </div>
  );
}

describe("Testing ThingContext matches snapshot", () => {
  it("matches snapshot with thing provided", () => {
    const property = "http://xmlns.com/foaf/0.1/name";
    const name = "example value";

    const exampleThing = SolidFns.addStringNoLocale(
      SolidFns.createThing(),
      property,
      name
    );

    documentBody = render(
      <ThingProvider thing={exampleThing}>
        <ExampleComponentWithThing />
      </ThingProvider>
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
  it("matches snapshot with thingUrl provided", () => {
    jest.spyOn(SolidFns, "getThingOne").mockImplementation(() => mockThing);

    documentBody = render(
      <DatasetProvider dataset={mockDataSetWithResourceInfo}>
        <ThingProvider thingUrl="https://some-interesting-value.com">
          <ExampleComponentWithThingUrl />
        </ThingProvider>
      </DatasetProvider>
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
});
