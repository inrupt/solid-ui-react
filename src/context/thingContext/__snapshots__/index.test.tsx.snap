// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ThingContext matches snapshot matches snapshot with thing provided 1`] = `
<body>
  <div>
    <div>
      <h2>
        example value
      </h2>
    </div>
  </div>
</body>
`;

exports[`Testing ThingContext matches snapshot matches snapshot with thingUrl provided 1`] = `
<body>
  <div>
    <div>
      <h2>
        test nick value
      </h2>
    </div>
  </div>
</body>
`;
